### ISSUE TITLE
Simple XOR Problem

### ISSUE THREAD

**Issue by bennean** [Sergio0694/NeuralNetwork.NET#96]
Thanks for the library, Sergio. I have a question that is similar to @quequiere's - I am trying to create a simple example (the XOR problem( to help understand your library but am struggling to get it to produce the correct results. Can you see what I'm doing incorrectly?

`	var network = NetworkManager.NewSequential(TensorInfo.Linear(2),
		NetworkLayers.FullyConnected(2, ActivationType.Sigmoid),
		NetworkLayers.Softmax(1));

	IEnumerable<(float[] x, float[] u)> data = new List<(float[] x, float[] u)>
	{
		(new[] {0f, 0f}, new[] {0f}),
		(new[] {0f, 1f}, new[] {1f}),
		(new[] {1f, 0f}, new[] {1f}),
		(new[] {1f, 1f}, new[] {0f})
	};

	var trainingData = DatasetLoader.Training(data, 10);

	var result = NetworkManager.TrainNetwork(
		network,
		trainingData,
		TrainingAlgorithms.StochasticGradientDescent(),
		100,
		0.5f,
		p => { Console.WriteLine($"Progress: {p.Percentage}%, Processed {p.ProcessedItems}"); },
		t => { Console.WriteLine($"Training: Iteration {t.Iteration}, Accuracy {t.Result.Accuracy}, Cost {t.Result.Cost}"); });

	Console.WriteLine(network.Forward(new float[] { 0, 0 })[0]);
	Console.WriteLine(network.Forward(new float[] { 0, 1 })[0]);
	Console.WriteLine(network.Forward(new float[] { 1, 0 })[0]);
	Console.WriteLine(network.Forward(new float[] { 1, 1 })[0]);`

**Comment by marseb7** [comment]
Hello,

Did you find the problem?

Best regards,

Sebastian

**Comment by bennean** [comment]
Sadly not!

A real shame because this framework seemed to show real promise. I ended up moving to Python & Keras.

Cheers,
Andy

**Comment by Sergio0694** [comment]
Hey, I'm sorry, but to be honest I haven't actually looked at this issue at all yet. The library is provided as is, and it's not a project I'm working on anymore. I believe there's a misunderstanding here regarding the scope of this project, since you mentioned Keras as an alternative - the two are not only not at the same level, they're not even in the same league at all.

Keras (TensorFlow in general) is arguably the world's most used neural network framework, it's used in production by hundreds of developers and researchers, maintained by thousands of engineers and sponsored by companies worth billions of dollars.

This on the other hand is literally just a small library I put together on my own in a couple months while at university, to learn how neural networks worked in detail and to do a few assignments in my university course. I've made it open source just because I thought I might've been interesting to look at for other developers, but this has never been and it never will a viable alternative to other well known neural network frameworks.

Hope this clears things up! 🙂

**Comment by PatriceDargenton** [comment]
Hello, I was able to successfully test your MNIST example, but I too failed with this simple XOR test, it would be great if you could provide a successful XOR example! Thank you.

**Comment by bennean** [comment]
Hi Sergio,

My comment was not intended to cause offence and I hope it wasn't taken as such. And I certainly wasn't comparing your library to Keras; it's just that the XOR problem is my ML 'Hello World' and I couldn't get it working so I moved on to another library, which happened to be Keras.

All the best,
Andy

**Comment by Sergio0694** [comment]
Hey @bennean - no worries, no offense was taken 😊

To be fair I'm not sure why the XOR example is not working, it could be anything from the library not working well with very small networks/datasets for some reason, or just something wrong with the learning settings, I just haven't tested that out myself. I've really only been using this library for MNIST and for similar image classification tasks with other datasets (eg. CIFAR and others), so I have personally never tried to setup and train very small network such as that, it might very well just be a bug there.

I've updated the readme though, in hindsight I think I should've made it clearer what the scope of the library was, and that I wasn't actively maintaining it anymore or personally investigating issues being open, so hopefully that'll help other devs as well.

Cheers!

**Comment by marseb7** [comment]
Hello,

Ok, thank you. What library would you recommend me for using in C# in small networks (150 hundreds neurons and 1 or 2 hidden layers)?

Best regards,

Sebastian

**Comment by Sergio0694** [comment]
The two other main libraries for C# that I know of are either [ML.NET](https://dotnet.microsoft.com/apps/machinelearning-ai/ml-dotnet) or [TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET), they should both work just fine. You can just pick the one you find easier to use (I believe that should be ML.NET as it's supposed to be a bit higher level I think).

That said at this point I might as well just see if I can get a XOR sample running with my library too because now I'm curious 🤣
I don't mean right now, but when I find the time I might get back to this and give it a try.

**Comment by PatriceDargenton** [comment]
@marseb7 I have tried several source codes and libraries with XOR there: https://github.com/PatriceDargenton/One-Ring-to-rule-them-all

**Comment by Wawrzyn321** [comment]
Hi, sorry to revive a closed thread, I know the project is not maintained anymore, but I added a [pull request](https://github.com/Sergio0694/NeuralNetwork.NET/pull/102) with solving the XOR problem as a sample project, feel free to take a look!

**Comment by bennean** [comment]
This looks really good, thank you!
